### DISCUSSION TITLE
How to use `extraRducers` pattern like redux tooltik?

### DISCUSSION THREAD

**Discussion by rwu823** [pmndrs/zustand#1654] [Q&A]
[Slice patterns](https://github.com/pmndrs/zustand/blob/main/docs/guides/slices-pattern.md) are helpful for maintaining large applications, and are similar to the redux toolkit's [createSlice](https://redux-toolkit.js.org/api/createslice), but how to use patterns like `extraReducers`? A further way is to treat state of slice as a state machine.

for example on doc:

```js
export const createFishSlice = (set) => ({
  fishes: 0,
  addFish: () => set((state) => ({ fishes: state.fishes + 1 })),
})

export const createBearSlice = (set) => ({
  bears: 0,
  addBear: () => set((state) => ({ bears: state.bears + 1 })),
  eatFish: () => set((state) => ({ fishes: state.fishes - 1 })),
})
```

But in the concepts of the state machine, the bear slice shouldn't care about how to mutate `fishes`

So it may become such:
```js
export const createBearSlice = (set) => ({
  bears: 0,
  addBear: () => set((state) => ({ bears: state.bears + 1 })),
  eatSomething: (type) => {
    // do something
  },
})

export const createFishSlice = (set) => ({
  fishes: 0,
  addFish: () => set((state) => ({ fishes: state.fishes + 1 })),
})

// This is just an example of how to rely on other actions of specify slice
createFishSlice.extraActions((slice) => {
  slice.addCase(createBearSlice().eatSomething, (state, payload: { type }) => {
    if (type === 'fish') {
      return { fishes: state.fishes - 1 }
    }
  })
})
```

Another proposal is that we can create the same action name, which will be executed on both sides.
(Now it will be overridden to execute only the last action in `create`)

```js
export const createBearSlice = (set) => ({
  bears: 0,
  addBear: () => set((state) => ({ bears: state.bears + 1 })),
  eatSomething: (type) => {
    // do something
  },
})

export const createFishSlice = (set) => ({
  fishes: 0,
  addFish: () => set((state) => ({ fishes: state.fishes + 1 })),
  eatSomething: (type) => set((state) => {
    if (type === 'fish) {
      return { fishes: state.fishes - 1 })
    }
  },
})
```

**Comment by dai-shi** [comment]
I don't think nobody created such middleware for your pattern. You would be the first to create one. 😄 

If it's basically injecting an action dynamically, what you could do in bare zustand is something like this:
```js
const useFooStore = ((set) => ({
  foo: 1,
  doSomething: () => { /* ... */ },
})
useFooStore.setState({ doSomething: () => { /* ... * } })
```

I don't think zustand slice pattern is same as RTK's. If you want namespacing, reach out some third-party libraries.

**Reply by oliwierptak** [reply]
I think the issue here is about inverting dependencies, and coming from RTK I'm not sure how to solve it.

```ts
// This is just an example of how to rely on other actions of specify slice
createFishSlice.extraActions((slice) => {
  slice.addCase(createBearSlice().eatSomething, (state, payload: { type }) => {
    if (type === 'fish') {
      return { fishes: state.fishes - 1 }
    }
  })
})
```

```ts
createFishSlice.extraActions((slice) => {
  slice.addCase(createBearSlice().eatSomething, (state, payload: { type }) => {
    if (type === 'apple') {
      return { apples: state.apples - 1 }
    }
  })
})
```

In the example above, it's the Fish and Apple slices that extend Bear slice functionality, while the Bear slice stays unmodified. 
In practice, it means that the Bear slice is probably in another package and does not have to be re-released when Bear will one day decie to eat berries
